pkg/migration-tool/planner/planner.go

```diff
@@ -0,0 +1,239 @@
+		}
+	}
+	plan.nextMint = plan.Mint
+	if plan.Mint >= plan.Maxt && found {
+		log.Info("msg", "mint greater than or equal to maxt. Migration has already been carried out.")
+		return false, nil
+	} else if plan.Mint >= plan.Maxt && !found {
+		// Extra sanitary check, even though this will be caught by validateConf().
+		return false, fmt.Errorf("mint cannot be greater than maxt: mint %d maxt %d", plan.Mint, plan.Mint)
+	}
+	plan.deltaIncRegion = plan.BlockSizeLimitBytes / 2 // 50% of the total block size limit.
+	return true, nil
+}
+
+// fetchLastPushedMaxt fetches the maxt of the last block pushed to remote-write storage. At present, this is developed
+// for a single migration job (i.e., not supporting multiple migration metrics and successive migrations).
+func (p *Plan) fetchLastPushedMaxt() (lastPushedMaxt int64, found bool, err error) {
+	query, err := utils.CreatePrombQuery(p.Mint, p.Maxt, []*labels.Matcher{
+		labels.MustNewMatcher(labels.MatchEqual, labels.MetricName, p.Config.ProgressMetricName),
+		labels.MustNewMatcher(labels.MatchEqual, utils.LabelJob, p.Config.JobName),
+	})
+	if err != nil {
+		return -1, false, fmt.Errorf("fetch-last-pushed-maxt create promb query: %w", err)
+	}
+	readClient, err := utils.NewClient("reader-last-maxt-pushed", p.Config.ProgressMetricURL, utils.Write, model.Duration(time.Minute*2))
+	if err != nil {
+		return -1, false, fmt.Errorf("create fetch-last-pushed-maxt reader: %w", err)
+	}
+	result, _, _, err := readClient.Read(context.Background(), query)
+	if err != nil {
+		return -1, false, fmt.Errorf("fetch-last-pushed-maxt query result: %w", err)
+	}
+	ts := result.Timeseries
+	if len(ts) == 0 {
+		return -1, false, nil
+	}
+	for _, series := range ts {
+		for i := len(series.Samples) - 1; i >= 0; i-- {
+			if series.Samples[i].Timestamp > lastPushedMaxt {
+				lastPushedMaxt = series.Samples[i].Timestamp
+			}
+		}
+	}
+	if lastPushedMaxt == 0 {
+		return -1, false, nil
+	}
+	return lastPushedMaxt, true, nil
+}
+
+func (p *Plan) DecrementBlockCount() {
+	p.blockCounts.Sub(1)
+}
+
+// ShouldProceed reports whether the fetching process should proceeds further. If any time-range is left to be
+// fetched from the provided time-range, it returns true, else false.
+func (p *Plan) ShouldProceed() bool {
+	return p.nextMint < p.Maxt
+}
+
+// update updates the details of the planner that are dependent on previous fetch stats.
+func (p *Plan) update(numBytes int) {
+	p.lastNumBytes.Store(int64(numBytes))
+}
+
+func (p *Plan) LastMemoryFootprint() int64 {
+	return p.lastNumBytes.Load()
+}
+
+// NewBlock returns a new block after allocating the time-range for fetch.
+func (p *Plan) NextBlock() (reference *Block, err error) {
+	timeDelta := determineTimeDelta(p.lastNumBytes.Load(), p.BlockSizeLimitBytes, p.deltaIncRegion, p.lastTimeRangeDelta)
+	mint := p.nextMint
+	maxt := mint + timeDelta
+	if maxt > p.Maxt {
+		maxt = p.Maxt
+	}
+	p.nextMint = maxt + 1
```

Comment:
I don't think there should be a +1 here, if maxt is excusive

Reply:
I think the reason for this is than, to progress the next block timestamps. If I don't add +1, then it wil not proceed.